app/controllers/subsidiaries/GroupSchemeController.scala

```diff
@@ -40,39 +41,77 @@ class GroupSchemeController @Inject()(val mcc: MessagesControllerComponents,
                                       implicit val countryCodes: CountryCodes,
                                       implicit val ersUtil: ERSUtil,
                                       implicit val sessionService: FrontendSessionService,
+                                      implicit val companyService: CompanyDetailsService,
                                       implicit val appConfig: ApplicationConfig,
                                       globalErrorView: views.html.global_error,
                                       groupView: views.html.group,
                                       groupPlanSummaryView: views.html.group_plan_summary,
+                                      confirmDeleteCompanyView: views.html.confirm_delete_company,
+                                      yesNoFormProvider: YesNoFormProvider,
                                       authAction: AuthAction
                                      ) extends FrontendController(mcc) with I18nSupport with WithUnsafeDefaultFormBinding with Logging with Constants with CacheHelper {
 
   implicit val ec: ExecutionContext = mcc.executionContext
 
-  def deleteCompany(id: Int): Action[AnyContent] = authAction.async {
-    implicit request =>
-      showDeleteCompany(id)(request)
+  private val form: Form[Boolean] = yesNoFormProvider.withPrefix("delete-company")
+
+  def confirmDeleteCompanyPage(id: Int): Action[AnyContent] = authAction.async { implicit request =>
+    showConfirmDeleteCompanyPage(id)
   }
 
-  def showDeleteCompany(id: Int)(implicit request: RequestWithOptionalAuthContext[AnyContent]): Future[Result] = {
+  def showConfirmDeleteCompanyPage(id: Int)(implicit request: RequestWithOptionalAuthContext[AnyContent]): Future[Result] =
     (for {
-      all <- sessionService.fetchAll()
-      companies = getEntry[CompanyDetailsList](all, DataKey(ersUtil.SUBSIDIARY_COMPANIES_CACHE)).getOrElse(CompanyDetailsList(Nil))
-      companyDetailsList = CompanyDetailsList(filterDeletedCompany(companies, id))
-      _ <- sessionService.cache(ersUtil.SUBSIDIARY_COMPANIES_CACHE, companyDetailsList)
+      requestObject   <- sessionService.fetch[RequestObject](ersUtil.ERS_REQUEST_OBJECT)
```

Comment:
Only problem is we'd still need to handle the future so would end up with a repeated transform in the method anyway still leading to code duplication. I think it might just be better to be left honestly, I can't really think of a clean way of handling it all

Reply:
Actually nevermind, could do something like:

``` scala
    getRequestObjAndCompanyDetails() transformWith {
      case res @ Success(_) => Future(Ok(confirmDeleteCompanyView(res.value._1, id, form, res.value._2 == 1, res.value._3.companies(id).companyName)))
      case Failure(cause) => Future(getGlobalErrorPage)
    }
```

Reply:
Though the result values definitely hurt readability

Reply:
```scala
  def showConfirmDeleteCompanyPage(id: Int)(implicit request: RequestWithOptionalAuthContext[AnyContent]): Future[Result] = {
    getRequestObjAndCompanyDetails() transformWith {
      case res @ Success(_) =>
        val requestObject = res.value._1
        val companyLength = res.value._2
        val companyDetailsList = res.value._3
        Future(Ok(confirmDeleteCompanyView(requestObject, id, form, companyLength == 1, companyDetailsList.companies(id).companyName)))
      case Failure(cause) =>
        logger.error(
          s"[GroupSchemeController][showConfirmDeleteCompany] getRequestObjAndCompanyDetails failed, timestamp: ${System.currentTimeMillis()}, error: $cause"
        )
        Future(getGlobalErrorPage)
    }
```

Maybe something like this on the method to call? what do you think

Reply:
Or this:

``` scala
  def showConfirmDeleteCompanyPage(id: Int)(implicit request: RequestWithOptionalAuthContext[AnyContent]): Future[Result] = {
    getRequestObjAndCompanyDetails() transformWith {
      case _ @ Success((requestObject: RequestObject, companySize: Int, companyDetailsList: CompanyDetailsList)) =>
        Future(Ok(confirmDeleteCompanyView(requestObject, id, form, companySize == 1, companyDetailsList.companies(id).companyName)))
      case Failure(cause) =>
        logger.error(
          s"[GroupSchemeController][showConfirmDeleteCompany] getRequestObjAndCompanyDetails failed, timestamp: ${System.currentTimeMillis()}, error: $cause"
        )
        Future(getGlobalErrorPage)
    } recover { case _: Exception =>
      getGlobalErrorPage
    }
  }
```